lib/chef/resource/inspec_waiver_file.rb

```diff
@@ -0,0 +1,152 @@
+#
+# Copyright:: Copyright (c) Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "../resource"
+require_relative "../dist"
+require "yaml" unless defined?(YAML)
+
+class Chef
+  class Resource
+    class InspecWaiverFile < Chef::Resource
+      provides :inspec_waiver_file
+      unified_mode true
+
+      description "Use the **inspec_waiver_file** resource to add or remove entries from an inspec waiver file. This can be used in conjunction with the audit cookbook"
+      introduced "16.3"
+      examples <<~DOC
+      **Add an InSpec waiver entry to a given waiver file**:
+
+      ```ruby
+        inspec_waiver_file 'Add waiver entry for control' do
+          file 'C:\\chef\\inspec_waiver.yml'
+          control 'my_inspec_control_01'
+          run_test false
+          justification "The subject of this control is not managed by Chef on the systems in policy group \#{node['policy_group']}"
+          expiration '2021-01-01'
+          action :add
+        end
+      ```
+
+      **Add an InSpec waiver entry to a given waiver file using the 'name' property to identify the control**:
+
+      ```ruby
+        inspec_waiver_file 'my_inspec_control_01' do
+          file 'C:\\chef\\inspec_waiver.yml'
+          justification "The subject of this control is not managed by Chef on the systems in policy group \#{node['policy_group']}"
+          action :add
+        end
+      ```
+
+      **Remove an InSpec waiver entry to a given waiver file**:
+
+      ```ruby
+        inspec_waiver_file "my_inspec_control_01" do
+          file '/etc/chef/inspec_waiver.yml'
+          action :remove
+        end
+      ```
+      DOC
+
+      property :control, String,
+        name_property: true,
+        description: "The name of the control being added or removed to the waiver file"
+
+      property :file, String,
+        required: true,
+        description: "The path to the waiver file being modified"
+
+      property :expiration, String,
+        description: "The expiration date of the given waiver - provided in YYYY-MM-DD format",
```

Comment:
Using a regex to validate the form of the string makes sense, but instead of validating the content (eg, that the third stanza can be 30 or 31 but not 32), this could rely on Date parsing:

```
irb(main):001:0> require 'date'
=> true
irb(main):002:0> s = "2020-01-31"
=> "2020-01-31"
irb(main):003:0> Date.valid_date?(*s.split("-").map(&:to_i))
=> true
irb(main):004:0> s = "2020-02-31"
=> "2020-02-31"
irb(main):005:0> Date.valid_date?(*s.split("-").map(&:to_i))
=> false
```